password_enhancements.module

```diff
@@ -0,0 +1,416 @@
+  password_enhancements_form_user_register_form_alter($form, $form_state, $form_id);
+  $form['password_change_required']['#access'] = Drupal::currentUser()->hasPermission('administer user password settings');
+  array_unshift($form['actions']['submit']['#submit'], 'password_enhancements_form_user_form_alter_submit');
+}
+
+/**
+ * Submit callback for password_enhancements_form_user_form_alter().
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ */
+function password_enhancements_form_user_form_alter_submit(array &$form, FormStateInterface $form_state): void {
+  // Update last updated date for the password.
+  if (!empty($form_state->getValue('pass'))) {
+    $form_state->setValue('password_changed', Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_password_enhancements_change_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  password_enhancements_form_user_register_form_alter($form, $form_state, $form_id);
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_user_register_form_alter(array &$form, FormStateInterface $form_state, string $form_id): void {
+  /** @var \Drupal\Core\Config\Config $config */
+  $config = Drupal::getContainer()->get('config.factory')->get('password_enhancements.settings');
+  if (empty($form['account']['pass']) && !empty($config->get('require_password'))) {
+    $form['account']['pass'] = [
+      '#type' => 'password_confirm',
+      '#size' => 25,
+      '#description' => t('Provide a password for the new account in both fields.'),
+      '#required' => TRUE,
+    ];
+    $form['actions']['submit']['#submit'][] = 'password_enhancements_form_user_register_form_alter_password_submit';
+  }
+  elseif (empty($form['account']['pass'])) {
+    // No password field available, nothing to do here.
+    return;
+  }
+
+  // Add weight for account fields.
+  $weight = 0;
+  foreach ($form['account'] as &$field) {
+    if (is_array($field) && !array_key_exists('#weight', $field)) {
+      $field['#weight'] = $weight++;
+    }
+  }
+
+  // Get roles.
+  $roles = $form_state->getValue('roles') ?? [];
+  $input = $form_state->getUserInput();
+  // If there was no input then grab the default roles.
+  if (empty($roles) && empty($input['roles']) && !empty($form['account']['roles']['#default_value'])) {
+    $roles = $form['account']['roles']['#default_value'];
+  }
+  $roles = array_merge(['authenticated'], $roles, (!empty($input['roles']) ? array_values($input['roles']) : []));
+
+  $constraints = _password_enhancements_get_constraints($roles);
+  $processed_constraints = [];
+  $descriptions = [];
+  $libraries = [
+    'password_enhancements/requirements',
+  ];
+  $plugin_config = [];
+  foreach ($constraints as $id => $constraint) {
```

Comment:
What if you would move this constraint discovery to a new function which only gets called here? That would reduce the complexity of this function and it could be tested individually if needed?

Reply:
I think if this part ever needs to be used somewhere else then it would make sense to create a separate function for it, but until then I wouldn't separate it from here.

Reply:
Well, re-usability is not only the mean reason why a developer should explode a function to smaller components :) Maintainability, readability, and testability are also important. This function is 113 lines at this moment.

Reply:
I don't see any reason to explode this part to another function, also it has 4 different output that I need (ok, I could return them as an array, but why?). It doesn't make it any harder to test it as it is, because the return of the form alter contains the outputs.

I would gladly create a function from this if it would return only one thing, or a list from the same type, but returning 4 different things from a function just have to original function shorter would just add more confusion I think.

Reply:
Well if one function needs to produce 4 different output it should be either done by defining a class for the output or by exploding it to 4 smaller functions instead of one. Let's get back to this later.

Reply:
Okay, I think you won't change your mind so I'll mark it as resolved, because it leaves more space for other comments ;P :D